### DISCUSSION TITLE
skywalking agent show java.lang.ClassCastException: org.apache.skywalking.apm.plugin.vertx3.HttpClientRequestImplInterceptor$HttpClientRequestContext cannot be cast to org.apache.skywalking.apm.plugin.vertx3.VertxContext

### DISCUSSION THREAD

**Discussion by kingkong2019** [apache/skywalking#8454] [Q&A] [answered]
skywalking agent log show java.lang.ClassCastException: org.apache.skywalking.apm.plugin.vertx3.HttpClientRequestImplInterceptor$HttpClientRequestContext cannot be cast to org.apache.skywalking.apm.plugin.vertx3.VertxContext

ERROR 2022-01-13 10:01:53:906 vert.x-eventloop-thread-1 InstMethodsInter : class[class io.vertx.core.http.impl.HttpClientRequestImpl] before method[handleException] intercept failure 
java.lang.ClassCastException: org.apache.skywalking.apm.plugin.vertx3.HttpClientRequestImplInterceptor$HttpClientRequestContext cannot be cast to org.apache.skywalking.apm.plugin.vertx3.VertxContext
	at org.apache.skywalking.apm.plugin.vertx3.HttpClientRequestImplHandleExceptionInterceptor.beforeMethod(HttpClientRequestImplHandleExceptionInterceptor.java:33)
	at org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter.intercept(InstMethodsInter.java:76)
	at io.vertx.core.http.impl.HttpClientRequestImpl.handleException(HttpClientRequestImpl.java)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleException(Http1xClientConnection.java:525)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.reset(Http1xClientConnection.java:377)
	at io.vertx.core.http.impl.HttpClientRequestImpl.reset(HttpClientRequestImpl.java:294)
	at io.vertx.core.http.impl.HttpClientRequestBase.handleTimeout(HttpClientRequestBase.java:195)
	at io.vertx.core.http.impl.HttpClientRequestBase.lambda$setTimeout$0(HttpClientRequestBase.java:118)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:942)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.handle(VertxImpl.java:906)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:221)
	at io.vertx.core.impl.VertxImpl$InternalTimerHandler.run(VertxImpl.java:932)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:170)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:503)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)

**Comment by wu-sheng** [comment]
It seems you are using vertx in a different case comparing the plugin expected. You could cross-check with the existing cases, 3 tested cases in here, https://github.com/apache/skywalking-java/tree/main/test/plugin/scenarios

**Comment by kingkong2019** [comment]
Thanks for your reply.
The agent version is 8.8.0,and what I can do to fix this issue?

**Reply by wu-sheng** [reply]
We don't know how to fix. Every framework has countless use cases, someone contributed this plugin with the cases they used. That is all.
Once you faced uncovered cases, you have to read the codes, understand what happens, then submit pull requests to fix this.

This is how open source works.

**Comment by kingkong2019** [comment]
Is the vertx is necessary to start? Can I bypassing the vertx when start agent to work?

**Reply by wu-sheng** [reply]
Yes, remove the plugin from the folder. But I doubt this would be helpful. No RPC relative instrumentation, all traces will break.

**Comment by kingkong2019** [comment]
Where I can get the apm-vertx-core-3.x-plugin-8.8.0.jar source code?

**Reply by wu-sheng** [reply]
https://github.com/apache/skywalking-java/tree/main/apm-sniffer/apm-sdk-plugin/vertx-plugins

**Comment by kingkong2019** [comment]
I found some warning log，did it cause the error？

WARN 2022-01-21 08:06:51:873 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.oas.web.OpenApiControllerWebMvc by plugin org.apache.skywalking.apm.plugin.spring.mvc.v5.define.reactive.ReactiveRestControllerInstrumentation is not working. Because witness class org.springframework.web.reactive.result.method.InvocableHandlerMethod is not existed.
WARN 2022-01-21 08:06:51:875 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.oas.web.OpenApiControllerWebMvc by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:51:875 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.oas.web.OpenApiControllerWebMvc by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerForLowVersionInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:52:235 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger.web.ApiResourceController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v5.define.reactive.ReactiveRestControllerInstrumentation is not working. Because witness class org.springframework.web.reactive.result.method.InvocableHandlerMethod is not existed.
WARN 2022-01-21 08:06:52:235 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger.web.ApiResourceController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:52:235 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger.web.ApiResourceController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerForLowVersionInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:52:297 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger2.web.Swagger2ControllerWebMvc by plugin org.apache.skywalking.apm.plugin.spring.mvc.v5.define.reactive.ReactiveRestControllerInstrumentation is not working. Because witness class org.springframework.web.reactive.result.method.InvocableHandlerMethod is not existed.
WARN 2022-01-21 08:06:52:297 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger2.web.Swagger2ControllerWebMvc by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:52:297 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger2.web.Swagger2ControllerWebMvc by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerForLowVersionInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:53:303 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v3.define.ControllerInstrumentation is not working. Because witness class org.springframework.web.servlet.view.xslt.AbstractXsltView is not existed.
WARN 2022-01-21 08:06:53:304 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v5.define.reactive.ReactiveControllerInstrumentation is not working. Because witness class org.springframework.web.reactive.result.method.InvocableHandlerMethod is not existed.
WARN 2022-01-21 08:06:53:304 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.ControllerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:06:53:304 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.ControllerForLowVersionInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:07:08:510 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.web.method.HandlerMethod by plugin org.apache.skywalking.apm.plugin.spring.mvc.v3.define.HandlerMethodInstrumentation is not working. Because witness class org.springframework.web.servlet.view.xslt.AbstractXsltView is not existed.
WARN 2022-01-21 08:07:08:513 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.web.method.HandlerMethod by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.HandlerMethodInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:07:09:865 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger2.web.Swagger2ControllerWebFlux by plugin org.apache.skywalking.apm.plugin.spring.mvc.v5.define.reactive.ReactiveRestControllerInstrumentation is not working. Because witness class org.springframework.web.reactive.result.method.InvocableHandlerMethod is not existed.
WARN 2022-01-21 08:07:09:865 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger2.web.Swagger2ControllerWebFlux by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:07:09:865 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.swagger2.web.Swagger2ControllerWebFlux by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerForLowVersionInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:07:09:882 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.oas.web.OpenApiControllerWebFlux by plugin org.apache.skywalking.apm.plugin.spring.mvc.v5.define.reactive.ReactiveRestControllerInstrumentation is not working. Because witness class org.springframework.web.reactive.result.method.InvocableHandlerMethod is not existed.
WARN 2022-01-21 08:07:09:882 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.oas.web.OpenApiControllerWebFlux by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:07:09:882 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class springfox.documentation.oas.web.OpenApiControllerWebFlux by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.RestControllerForLowVersionInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.
WARN 2022-01-21 08:07:13:476 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.web.method.support.InvocableHandlerMethod by plugin org.apache.skywalking.apm.plugin.spring.mvc.v3.define.InvocableHandlerInstrumentation is not working. Because witness class org.springframework.web.servlet.view.xslt.AbstractXsltView is not existed.
WARN 2022-01-21 08:07:13:476 Catalina-utility-1 AbstractClassEnhancePluginDefine : enhance class org.springframework.web.method.support.InvocableHandlerMethod by plugin org.apache.skywalking.apm.plugin.spring.mvc.v4.define.InvocableHandlerInstrumentation is not working. Because witness class org.springframework.cache.interceptor.DefaultKeyGenerator is not existed.

**Reply by wu-sheng** [reply]
These warnings are fine. That only means some plugins are disabled automatically because you are not using relative frameworks.

**Comment by kingkong2019** [comment]
It's very strange thing: today I restart the service ,there is no error message,and the trace function is normal now.

**Reply by wu-sheng** [reply]
Sorry, plugins are various from global.  We don't know all details. But I am happy your side is working

**Comment by kingkong2019** [comment, marked as answer]
I fixed this issue,the problem code as below:

package org.apache.skywalking.apm.plugin.vertx3;
public class HttpClientRequestImplHandleExceptionInterceptor implements InstanceMethodsAroundInterceptor {

    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
        MethodInterceptResult result) throws Throwable {
        **VertxContext context = (VertxContext) objInst.getSkyWalkingDynamicField();**
        context.getSpan().log((Throwable) allArguments[0]);
    }

    changed as below:

    public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
        MethodInterceptResult result) throws Throwable {
        HttpClientRequestImplInterceptor.HttpClientRequestContext
                requestContext = (HttpClientRequestImplInterceptor.HttpClientRequestContext) objInst.getSkyWalkingDynamicField();
        VertxContext context = requestContext.vertxContext;
        context.getSpan().log((Throwable) allArguments[0]);
    }

**Reply by wu-sheng** [reply]
It is better for you to submit a pull request. Would you do that?

**Reply by kingkong2019** [reply]
Sure, after I fixed another issue about span leak I will do it together.
